[SYCL][Bindless][Doc] Introduce SPIR-V spec for enabling bindless images on Intel devices (#12927)
intel/llvm: sycl/doc/design/spirv-extensions/SPV_INTEL_bindless_images.asciidoc

SPV_INTEL_bindless_images
=========================

== Name Strings

SPV_INTEL_bindless_images

== Contact

To report problems with this extension, please open a new issue at:

<https://github.com/intel/llvm/issues>

== Contributors

- Duncan Brawley, Codeplay
- Przemek Malon, Codeplay
- Peter Žužek, Codeplay
- Chedy Najjar, Codeplay
- Sean Stirling, Codeplay
- Isaac Ault, Codeplay
- Victor Lomuller, Codeplay

== Notice

Copyright © Codeplay Software Limited. All rights reserved.

== Status

In Development

== Version

[width="40%",cols="25,25"]
|========================================
| Last Modified Date | 2024-02-23
| Revision           | 6
|========================================

== Dependencies

This extension is written against the SPIR-V Specification,
Version 1.6 Revision 1.

This extension requires SPIR-V 1.0.

== Overview

This extension adds support for bindless images.
This is done by adding support for SPIR-V to convert unsigned integer handles to images/samplers.

> Bindless images are a feature that provides flexibility on how images are accessed and used, such as removing limitations on how many images can be accessed as well as potentially improving performance.

Author: Ah good point. I had a poke around and I don't think there is a SPIR-V spec style guide? Though the convention seems to be to keep to the 80 length limit (at least mostly) so will stick to that.

This is an improvement on the legacy bound images model which is a holdover from binding slots in hardware which used to be limited in number.

== Extension Name

To use this extension within a SPIR-V module, the following *OpExtension* must be present in the module:

----
OpExtension "SPV_INTEL_bindless_images"
----

== New Capabilities

This extension introduces a new capability:

----
BindlessImagesINTEL
----

== New Instructions

Instructions added under *BindlessImagesINTEL* capability.

----
OpConvertHandleToImageINTEL
OpConvertHandleToSamplerINTEL
----

== Token Number Assignments

--
[width="40%"]
[cols="70%,30%"]
[grid="rows"]
|====
|BindlessImagesINTEL                |6528
|OpConvertHandleToImageINTEL        |6529
|OpConvertHandleToSamplerINTEL      |6530
|====
--

> == Modifications to the SPIR-V Specification, Version 1.6

Author: I have added a `modify section` for the new instructions. On stuff like capabilities, it seems pretty random if a SPIR-V spec has an explicit `modify section` for it. Though added it anyway as it feels more correct.

[cols="2*1,3*2"]
|======
5+|[[OpConvertHandleToImageINTEL]]*OpConvertHandleToImageINTEL* +
 +

> Accepts an unsigned integer and converts to image type.

Reviewer: ```suggestion
Converts an unsigned integer pointed by _Operand_ to image type.
```

Author: Applied.

Unsigned integer is either a 64 or 32 bit unsigned integer.

> Depending on whether *OpMemoryModel* is set to *Physical32* or *Physical64*.

Reviewer: It is probably better to say if "Addressing Model" is set to .... Even though Addressing Model is set through `OpMemoryModel`, there is also Memory Model "enum" as well and here we are specifically talking about addressing

Author: Updated wording to use `addressing model`.

> Return type must be an `OpTypeImage`.

Reviewer: ```suggestion
_Result type_ must be an `OpTypeImage`.
```

Author: Applied.

> | 4 | 6529 | '<id> Result Type' | 'Result <id>' | '<id> Operand'

Author: Interesting. Did not know you could do that. I have added them for the instruction numbers.
